src/vsphere_cpi/lib/cloud/vsphere/nsxt_provider.rb

```diff
@@ -118,42 +124,58 @@ def initialize(client, default_vif_type)
     def add_vm_to_nsgroups(vm, ns_groups)
       return if ns_groups.nil? || ns_groups.empty?
       return if nsxt_nics(vm).empty?
-
-      logger.info("Adding vm '#{vm.cid}' to NSGroups: #{ns_groups}")
-      nsgroups = retrieve_nsgroups(ns_groups)
-
       lports = logical_ports(vm)
-      nsgroups.each do |nsgroup|
-        logger.info("Adding LogicalPorts: #{lports.map(&:id)} to NSGroup '#{nsgroup.id}'")
-        grouping_obj_svc.add_or_remove_ns_group_expression(
-          nsgroup.id,
-          *to_simple_expressions(lports),
-          'ADD_MEMBERS'
-        )
+      logger.info("Adding vm '#{vm.cid}' to NSGroups: #{ns_groups}")
+      Bosh::Retryable.new(
+          tries: 5,
+          on: [NSXTSecGrpUpdateError]
+      ).retryer do |i|
+        nsgroups = retrieve_nsgroups(ns_groups)
+        nsgroups.each do |nsgroup|
+          logger.info("Adding LogicalPorts: #{lports.map(&:id)} to NSGroup '#{nsgroup.id}'")
+          begin
+            grouping_obj_svc.add_or_remove_ns_group_expression(
```

Comment:
I think there's a big problem here. If we hit HTTP 409/412 this code will *retry* all of the NSGroup modifications instead of the single one that failed. I'm not sure that `ADD_MEMBERS` is idempotent so we may end up adding the wrong members to the NSGroup.

Reply:
Moved `BOSH::Retryable` block inside the loop so that it retries on the single `ADD_MEMBERS` expression that failed in the current iteration.